### PR TITLE
OSSM-3979 Adding EnvoyExtAuthzGRPC to GA

### PR DESCRIPTION
**Pull request by mayleighnmyers**

### CODE REVIEW

**Review by jewertow — CHANGES_REQUESTED** [review]
Please compare `ExtensionProviderEnvoyExternalAuthorizationGRPCConfig` with this [structure](https://github.com/istio/api/blob/master/mesh/v1alpha1/config.pb.go#L2521-L2547) and remove fields which don't exist there.

File: pkg/apis/maistra/v2/extension_providers.go [hunk 1]
```diff
@@ -57,6 +60,32 @@ type ExtensionProviderEnvoyExternalAuthorizationHTTPConfig struct {
...
+	// The maximum duration that the proxy will wait for a response from the provider (default timeout: 600s).
+	// When this timeout condition is met, the proxy marks the communication to the authorization service as failure.
+	// In this situation, the response sent back to the client will depend on the configured `fail_open` field.
+	Timeout *string `json:"timeout,omitempty"`
+	// Sets a prefix to the value of authorization request header *Path*.
+	// For example, setting this to "/check" for an original user request at path "/admin" will cause the
+	// authorization check request to be sent to the authorization service at the path "/check/admin" instead of "/admin".	FailOpen *bool `json:"failOpen,omitempty"`
+	// Sets the HTTP status that is returned to the client when there is a network error to the authorization service.
+	// The default status is "403" (HTTP Forbidden).
+	FailOpen *bool `json:"fail_open,omitempty"`
```
Comment (jewertow) [inline pkg/apis/maistra/v2/extension_providers.go]: JSON schemas are camel case.
```suggestion
	FailOpen *bool `json:"failOpen,omitempty"`
```

**Review by jewertow — CHANGES_REQUESTED** [review]

**Comment by jewertow** [comment]
Could you rebase your branch?

File: build/patch-charts.sh
Comment (jewertow) [inline build/patch-charts.sh:1]: Seems like an unrelated change...

File: resources/helm/v2.4/gateway/templates/poddisruptionbudget.yaml
Comment (jewertow) [inline resources/helm/v2.4/gateway/templates/poddisruptionbudget.yaml:1]: Same here.

File: resources/helm/v2.4/gateway/values.schema.json
Comment (jewertow) [inline resources/helm/v2.4/gateway/values.schema.json:1]: Same here.

File: resources/helm/v2.4/gateway/values.yaml
Comment (jewertow) [inline resources/helm/v2.4/gateway/values.yaml:1]: Same here.

File: build/patch-charts.sh
Comment (mayleighnmyers) [reply build/patch-charts.sh:1]: I mean they aren't related to Grpc, however, my `make gen` wasn't running through the test so Marko helped and we found this code to be an issue. I could take it out and put it in a new PR

File: resources/helm/v2.4/gateway/templates/poddisruptionbudget.yaml
Comment (mayleighnmyers) [reply resources/helm/v2.4/gateway/templates/poddisruptionbudget.yaml:1]: This must've came into the PR when I ran `make gen` since I don't remember doing this.

File: resources/helm/v2.4/gateway/values.schema.json
Comment (mayleighnmyers) [reply resources/helm/v2.4/gateway/values.schema.json:1]: This must've came into the PR when I ran make gen since I don't remember doing this.

File: resources/helm/v2.4/gateway/values.yaml
Comment (mayleighnmyers) [reply resources/helm/v2.4/gateway/values.yaml:1]: This must've came into the PR when I ran make gen since I don't remember doing this.

File: resources/helm/v2.4/gateway/values.schema.json
Comment (luksa) [reply resources/helm/v2.4/gateway/values.schema.json:1]: This happened because some dependencies were upgrade but `make gen` wasn't run. You don't need to remove it from this PR, as it's already in #1257.  You could rebase once that PR is merged, but it's not necessary.

File: resources/helm/v2.4/gateway/values.yaml
Comment (luksa) [reply resources/helm/v2.4/gateway/values.yaml:1]: This is now part of #1257. No need to remove, but you could rebase when that PR is merged.

**Comment by jewertow** [comment]
@mayleighnmyers could you first rebase your branch and fix failed checks?

File: pkg/apis/maistra/v2/extension_providers.go [hunk 2]
```diff
@@ -52,20 +55,44 @@ type ExtensionProviderEnvoyExternalAuthorizationHTTPConfig struct {
...
+	// REQUIRED. Specifies the port of the service.
+	Port int64 `json:"port"`
+	// The maximum duration that the proxy will wait for a response from the provider, this is the timeout for a specific request (default timeout: 600s).
+	// When this timeout condition is met, the proxy marks the communication to the authorization service as failure.
+	// In this situation, the response sent back to the client will depend on the configured `failOpen` field.
+	Timeout *string `json:"timeout,omitempty"`
+	// If true, the HTTP request or TCP connection will be allowed even if the communication with the authorization service has failed,
+	// or if the authorization service has returned a HTTP 5xx error.
+	// Default is false. For HTTP request, it will be rejected with 403 (HTTP Forbidden). For TCP connection, it will be closed immediately.
+	FailOpen bool `json:"failOpen,omitempty"`
```
Comment (mayleighnmyers) [inline pkg/apis/maistra/v2/extension_providers.go]: I noticed there were pointers @jewertow had used, is this fine to not use a pointer (following upstream). Or is there a reason for us to change it to a pointer?

File: pkg/apis/maistra/v2/extension_providers.go [hunk 3]
```diff
@@ -52,20 +55,44 @@ type ExtensionProviderEnvoyExternalAuthorizationHTTPConfig struct {
...
+	// Sets the HTTP status that is returned to the client when there is a network error to the authorization service.
+	// The default status is "403" (HTTP Forbidden).
+	StatusOnError string `json:"statusOnError,omitempty"`
```
Comment (mayleighnmyers) [inline pkg/apis/maistra/v2/extension_providers.go]: Same here regarding the pointers

**Comment by mayleighnmyers** [comment]
> @mayleighnmyers could you first rebase your branch and fix failed checks?

I rebased, made a few changes, and ran all tests including `make gen` which shows the only changes being the time stamp

File: pkg/apis/maistra/conversion/extension_providers_test.go [hunk 4]
```diff
@@ -172,6 +172,46 @@ meshConfig:
...
+    envoyExtAuthzGrpc:
+      service: ext-authz.foo.svc.cluster.local
+      port: 8000
+      timeout: 1s
+      failOpen: true
+      statusOnError: "500"
+      includeRequestBodyInCheck:
+        maxRequestBytes: 100
+        allowPartialMessage: true
+        packAsBytes: true
```
Comment (luksa) [inline pkg/apis/maistra/conversion/extension_providers_test.go:215]: Are these field names correct? Is it `failOpen` or `fail_open`? Both formats appear in the upstream type, but I'm not sure which one is actually used:
```
FailOpen bool `protobuf:"varint,4,opt,name=fail_open,json=failOpen,proto3" json:"fail_open,omitempty"`
                                           ^^^^^^^^^      ^^^^^^^^               ^^^^^^^^^
```

Comment (mayleighnmyers) [reply pkg/apis/maistra/conversion/extension_providers_test.go:215]: We were following Jacek's work by doing the camel casing instead of snake casing. If we want to be more inline with upstream, I am more than happy to change! :)

Comment (luksa) [reply pkg/apis/maistra/conversion/extension_providers_test.go:215]: Yes, I know, but I'm talking specifically about the helmValues, not about the SMCP (where, as you say, we followed Jacek's work). For helmValues, we need to use whatever istio expects.

Comment (luksa) [reply pkg/apis/maistra/conversion/extension_providers_test.go:215]: Okay, it looks like it expects the fields to be in camel case, so we're good.

https://istio.io/latest/docs/reference/config/istio.mesh.v1alpha1/#MeshConfig-ExtensionProvider-EnvoyExternalAuthorizationGrpcProvider

Comment (jewertow) [reply pkg/apis/maistra/conversion/extension_providers_test.go:215]: @luksa we use camel case, because we only support JSON schema.

File: pkg/apis/maistra/conversion/extension_providers.go [hunk 5]
```diff
@@ -131,8 +160,46 @@ func populateExtensionProvidersConfig(in *v1.HelmValues, out *v2.ControlPlaneSpe
...
 	}
-
 	return nil
 }
+
+func convertEnvoyExtAuthzGrpc(rawExtAuthz interface{}, out *v2.ControlPlaneSpec, provider map[string]interface{}) {
+	if extAuthz, ok := rawExtAuthz.(map[string]interface{}); ok {
+		config := &v2.ExtensionProviderEnvoyExternalAuthorizationGRPCConfig{
+			Service: extAuthz["service"].(string),
+			Port:    extAuthz["port"].(int64),
```
Comment (jewertow) [inline pkg/apis/maistra/conversion/extension_providers.go]: We should check casting result to avoid crashes.

File: pkg/apis/maistra/conversion/extension_providers.go [hunk 6]
```diff
@@ -131,8 +160,46 @@ func populateExtensionProvidersConfig(in *v1.HelmValues, out *v2.ControlPlaneSpe
...
+		}
+		if rawTimeout, ok := extAuthz["timeout"]; ok {
+			config.Timeout = strPtr(rawTimeout.(string))
+		}
+		if rawFailOpen, ok := extAuthz["failOpen"]; ok {
+			config.FailOpen = rawFailOpen.(bool)
```
Comment (jewertow) [inline pkg/apis/maistra/conversion/extension_providers.go]: Same here.

File: pkg/apis/maistra/conversion/extension_providers.go (hunk 5 above)
Comment (luksa) [reply pkg/apis/maistra/conversion/extension_providers.go]: Given that we need to fix these casts for ExtAuthzHTTP as well, I suggest we do it in a different PR.

File: pkg/apis/maistra/v2/extension_providers.go (hunk 2 above)
Comment (jewertow) [reply pkg/apis/maistra/v2/extension_providers.go]: We use pointers in API structures to know whether a particular field was specified by a user or not. In case of value fields, you don't know whether `failOpen: false` comes from user input or is it just the default bool value.
However, we only need to know that for required fields. Optional fields don't have to be pointers.

File: pkg/apis/maistra/conversion/extension_providers.go (hunk 5 above)
Comment (jewertow) [reply pkg/apis/maistra/conversion/extension_providers.go]: I was also thinking about it, but if we decide for that, we cannot release this change without the follow-up change with safe casting.

File: pkg/apis/maistra/v2/extension_providers.go (hunk 2 above)
Comment (jewertow) [reply pkg/apis/maistra/v2/extension_providers.go]: I was wrong, I'm sorry. We use pointers for optional values to know whether we should set them in helm values or not.

Comment (jewertow) [reply pkg/apis/maistra/v2/extension_providers.go]: But now I see that I didn't check default values in `validateExtensionProviders` for required fields, i.e. service and port. We can proceed this as is and improve that in a follow-up PR.

Comment (jewertow) [reply pkg/apis/maistra/v2/extension_providers.go]: So the following fields should be pointers: timeout, failOpen, statusOnError, includeRequestBodyInCheck.

Comment (jewertow) [reply pkg/apis/maistra/v2/extension_providers.go]: > to know whether we should set them in helm values or not.

Just for clarification: of course, we might use value fields and then set default values in helm templates, but then we would have to make sure that the default struct values are default API values. For example, default value for boolean field in Go is false, but an API might be implemented so that when it's not specified, then it's set to true.

File: pkg/apis/maistra/conversion/extension_providers.go (hunk 5 above)
Comment (mayleighnmyers) [reply pkg/apis/maistra/conversion/extension_providers.go]: Created a separate Jira to check casting https://issues.redhat.com/browse/OSSM-4628

File: pkg/apis/maistra/v2/extension_providers.go (hunk 2 above)
Comment (mayleighnmyers) [reply pkg/apis/maistra/v2/extension_providers.go]: I changed failOpen and statusOnError back to pointers. includeRequestBodyInCheck was already one.

**Comment by mayleighnmyers** [comment]
Marko's PR https://github.com/maistra/istio-operator/pull/1264, needs to be merged first before merging this PR.

File: pkg/controller/versions/strategy_v2_4.go [hunk 7]
```diff
@@ -203,13 +203,13 @@ func (v *versionStrategyV2_4) validateExtensionProviders(spec *v2.ControlPlaneSp
...
 	}
 	for _, ext := range spec.MeshConfig.ExtensionProviders {
-		if ext.Prometheus == nil && ext.EnvoyExtAuthzHTTP == nil {
+		if ext.Prometheus == nil && ext.EnvoyExtAuthzHTTP == nil && ext.EnvoyExtAuthzGRPC == nil {
 			allErrors = append(allErrors, fmt.Errorf("extension provider '%s' does not define any provider - "+
-				"it must specify one of: prometheus or envoyExtAuthzHttp", ext.Name))
+				"it must specify one of: prometheus, envoyExtAuthzHttp, or envoyExtAuthzGrcp", ext.Name))
 		}
-		if ext.Prometheus != nil && ext.EnvoyExtAuthzHTTP != nil {
+		if ext.Prometheus != nil && ext.EnvoyExtAuthzHTTP != nil && ext.EnvoyExtAuthzGRPC == nil {
```
Comment (luksa) [inline pkg/controller/versions/strategy_v2_4.go]: This if statement doesn't check what it's supposed to check. Maybe you should count the number of fields that are set and then rewrite this and the previous statement so they simply check if `count == 0` or `count > 1`.

File: pkg/controller/versions/strategy_v2_4.go [hunk 8]
```diff
@@ -202,18 +202,31 @@ func (v *versionStrategyV2_4) validateExtensionProviders(spec *v2.ControlPlaneSp
...
-		if ext.Prometheus == nil && ext.EnvoyExtAuthzHTTP == nil && ext.EnvoyExtAuthzGRPC == nil {
-			allErrors = append(allErrors, fmt.Errorf("extension provider '%s' does not define any provider - "+
-				"it must specify one of: prometheus, envoyExtAuthzHttp, or envoyExtAuthzGrcp", ext.Name))
-		}
-		if ext.Prometheus != nil && ext.EnvoyExtAuthzHTTP != nil && ext.EnvoyExtAuthzGRPC == nil {
-			allErrors = append(allErrors, fmt.Errorf("extension provider '%s' must specify only one type of provider: "+
-				"prometheus, envoyExtAuthzHttp, envoyExtAuthzGrpc", ext.Name))
-		}
+
+		counter := 0
```
Comment (luksa) [inline pkg/controller/versions/strategy_v2_4.go]: Better to move this to just before the if statements that do the counting (i.e. after the if statement that checks if the name is empty).

File: pkg/controller/versions/strategy_v2_4.go [hunk 9]
```diff
@@ -202,18 +202,31 @@ func (v *versionStrategyV2_4) validateExtensionProviders(spec *v2.ControlPlaneSp
...
+			counter++
+		}
+		if ext.EnvoyExtAuthzGRPC != nil {
+			counter++
+		}
+		if counter == 0 {
+			allErrors = append(allErrors, fmt.Errorf("extension provider '%s' does not define any provider - "+
+				"it must specify one of: prometheus, envoyExtAuthzHttp, or envoyExtAuthzGrpc", ext.Name))
+		} else {
+			if counter > 1 {
```
Comment (luksa) [inline pkg/controller/versions/strategy_v2_4.go]: This should be an `else if` instead of `else { if`.

File: pkg/controller/versions/strategy_v2_4.go [hunk 10]
```diff
@@ -222,15 +235,8 @@ func (v *versionStrategyV2_4) validateExtensionProviders(spec *v2.ControlPlaneSp
...
 			}
 		}
-		if ext.EnvoyExtAuthzGRPC != nil {
-			if ext.EnvoyExtAuthzGRPC.Timeout != nil {
-				if _, err := time.ParseDuration(*ext.EnvoyExtAuthzGRPC.Timeout); err != nil {
-					allErrors = append(allErrors, fmt.Errorf("invalid extension provider '%s': envoyExtAuthzGrcp.timeout "+
-						"must be specified in the duration format - got '%s'", ext.Name, *ext.EnvoyExtAuthzGRPC.Timeout))
-				}
-			}
-		}
```
Comment (luksa) [inline pkg/controller/versions/strategy_v2_4.go]: Why delete this?

File: pkg/apis/maistra/conversion/extension_providers.go [hunk 11]
```diff
@@ -62,6 +62,40 @@ func populateExtensionProvidersValues(in *v2.ControlPlaneSpec, allValues map[str
...
+					includeRequestBodyInCheckValues["maxRequestBytes"] = *config.IncludeRequestBodyInCheck.MaxRequestBytes
+				}
+				if config.IncludeRequestBodyInCheck.AllowPartialMessage != nil {
+					includeRequestBodyInCheckValues["allowPartialMessage"] = *config.IncludeRequestBodyInCheck.AllowPartialMessage
+				}
+				if config.IncludeRequestBodyInCheck.PackAsBytes != nil {
+					includeRequestBodyInCheckValues["packAsBytes"] = *config.IncludeRequestBodyInCheck.PackAsBytes
+				}
+				values["includeRequestBodyInCheck"] = includeRequestBodyInCheckValues
+			}
```
Comment (luksa) [inline pkg/apis/maistra/conversion/extension_providers.go]: This is duplicated code. Can you extract it into a function and call it for both HTTP and GRPC?

File: manifests-maistra/2.4.2/servicemeshcontrolplanes.crd.yaml
Comment (luksa) [inline manifests-maistra/2.4.2/servicemeshcontrolplanes.crd.yaml:1]: I believe you need to remove this file. This change will go into 2.4.3, so the only files that should be changed are the ones in the 2.4.3 directory.

File: manifests-servicemesh/2.4.2/servicemeshcontrolplanes.crd.yaml
Comment (luksa) [inline manifests-servicemesh/2.4.2/servicemeshcontrolplanes.crd.yaml:1]: Remove this file too.

File: pkg/apis/maistra/conversion/extension_providers.go [hunk 12]
```diff
@@ -44,31 +44,57 @@ func populateExtensionProvidersValues(in *v2.ControlPlaneSpec, allValues map[str
...
+			if config.Timeout != nil {
+				values["timeout"] = *config.Timeout
+			}
+			if config.FailOpen != nil {
+				values["failOpen"] = *config.FailOpen
+			}
+			if config.StatusOnError != nil {
+				values["statusOnError"] = *config.StatusOnError
+			}
+
```
Comment (luksa) [inline pkg/apis/maistra/conversion/extension_providers.go]: nit: please remove this line (and optionally add a blank line after lines 69 and 47).

**Review by luksa — APPROVED** [review]

**Review by jewertow — APPROVED** [review]

**Comment by tareqmamari** [comment]
@jewertow @mayleighnmyers @luksa we have a small question, is there any plans to release this change soon?

**Comment by mayleighnmyers** [comment]
> @jewertow @mayleighnmyers @luksa we have a small question, is there any plans to release this change soon?

@tareqmamari This will be release in the 2.4.3 release coming soon (a few days)
